R/result.R

```diff
@@ -1,36 +1,285 @@
+    },
+    read_json = function(path) {
+      self$from_json(paste(readLines(path), collapse = "\n"))
+    },
+    write_json = function(path) {
+      writeLines(self$json, path)
+    }
+  )
+)
+
+#' @export
+as.list.Serializable <- function(x, ...) x$list
+#' @export
+as.character.Serializable <- function(x, ...) x$json
+
+
+BenchmarkResult <- R6.1Class(
+  classname = "BenchmarkResult",
+  inherit = Serializable,
+
+  public = list(
+    initialize = function(name,
+                          result,
+                          params,
+                          tags = NULL,
+                          info = NULL,
+                          context = NULL,
+                          github = NULL,
+                          options = NULL,
+                          output = NULL,
+                          rscript = NULL) {
+      self$name <- name
+      self$result <- result
+      self$params <- params
+      self$tags <- tags
+      self$info <- info
+      self$context <- context
+      self$github <- github
+      self$options <- options
+      self$output <- output
+      self$rscript <- rscript
+    },
+
+    data.frame = function(row.names = NULL, optional = FALSE, packages = "arrow", ...) {
+      x <- self$list
+
+      pkgs <- x$params$packages
+      x$params$packages <- NULL
+      for (p in packages) {
+        x$params[[paste0("version_", p)]] <- pkgs[p, "version"]
+      }
+
+      to_list_col <- lengths(x$params) != 1L
+      x$params[to_list_col] <- lapply(x$params[to_list_col], list)
+
+      out <- dplyr::bind_cols(
+        iteration = seq_len(nrow(x$result)),
+        x$result,
+        dplyr::as_tibble(x$params)
+      )
+      out$output <- x$output
+
+      # append metadata fields to dataframe as attributes
+      metadata_elements <- c("name", "tags", "info", "context", "github", "options")
+      for (element in metadata_elements) {
+        if (element %in% names(x)) {
+          attr(out, element) <- x[[element]]
+        }
+      }
+
+      out
+    }
+  ),
+
+  active = list(
+    name = function(name) private$get_or_set(variable = ".name", value = name),
+    result = function(result) private$get_or_set(variable = ".result", value = result),
+    params = function(params) private$get_or_set(variable = ".params", value = params),
+    tags = function(tags) private$get_or_set(variable = ".tags", value = tags),
+    info = function(info) private$get_or_set(variable = ".info", value = info),
+    context = function(context) private$get_or_set(variable = ".context", value = context),
+    github = function(github) private$get_or_set(variable = ".github", value = github),
+    options = function(options) private$get_or_set(variable = ".options", value = options),
+    output = function(output) private$get_or_set(variable = ".output", value = output),
+    rscript = function(rscript) private$get_or_set(variable = ".rscript", value = rscript),
```

Comment:
Sort of unrelated to this but for validation will our best approach be to define validation functions a la:
```r
    name = function(name) is_ab_name(private$get_or_set(variable = ".name", value = name)),
    result = function(result) is_ab_result(private$get_or_set(variable = ".result", value = result))
    ...
```
Question more for my understanding than anything.

Reply:
Nope, we can't do that because the validation really needs to happen in the middle of that function, before it assigns with `private$.name <- name`. So probably we'll need to revert it to 

``` r
name = function(name) {
  if (!missing(name)) {
    # validation starts here, e.g.
    stopifnot(is_ab_name(name))
    # validation ends here
    private$.name <- name
  }
  private$.name
},
```

I thought about making `get_or_set()` take an expression or function for validation, but I don't want it to live in yet another location, so if we go that route, it would need to be defined in the call, e.g.

``` r
name = function(name) private$get_or_set(
  variable = ".name", 
  value = name, 
  validation = {stopifnot(is_ab_name(name))}
),
```
or something like that. But currently I prefer the more verbose version without `get_or_set()` because it's more transparent.